Integrate Sentry.io with existing back end logging

dzx11 opened the pull request:
Closes #130
-Uncomment the line in common.py/log_message to send messages to Sentry, however, be careful with this and avoid running too many tests with this on because we can easily go over the 5k events limit in Sentry. (We will probably need to add environment variables to control this and make sure it only runs in production. I haven't seen a way to do it in Sentry but it can be done with Python code).
-I haven't found a good way to test some of the code, particularly in game.py. It should be safe and error free, but the main concern is making sure that all of the game attributes possible are sent to log_message and that the data is in the right format.

Copy/pasting some info on the attributes in my email:

> Am I right in thinking that the logging function should be stateless? The reason that I'm asking is because a lot of attributes like game_id, player_id, phone_number, etc. are not available every time the logging function is called. They are only available in certain contexts. 
> 
> If we could save the state so that these attributes are saved the first time they are available, then it solves that problem, but from my understanding of the backend, the values would end up being incorrect and not accurately reflect the values of the user.
> 
> IF we can't save the state, many times the logging function would not have those values, and I'm not sure how much value logging those messages would bring, because we wouldn't be able to easily pair the logging message with the user who caused the message. Possibly we might be able to figure it out from location or browser, but only for a very small number of users.
>

dzx11 committed:
Add test logging to sentry

dzx11 committed:
Add game_id tag to sentry call

dzx11 committed:
Update senry project url

dzx11 committed:
Add phone_numbers, game_id, and player_id to logging and implement in to_dict

dzx11 committed:
Add support for more attributes in logging for test_events

dzx11 committed:
Refactor logging and update engine and events for it

dzx11 committed:
Update all log_message calls to include game attributes

dzx11 committed:
Add comments

dzx11 committed:
Merge branch 'develop' into feature/sentry

bernardbaker commented:
Hi David,

Reach out to the Sentry community and ask those questions about working
with a Python app which is in GCP and Netlify serverless functions which
are used by the front end.

Regards,

Bernard T. A. Baker


On Wed, May 13, 2020 at 3:58 PM dzx11 <notifications@github.com> wrote:

> Closes #130 <https://github.com/un7c0rn/stopthevirus/issues/130>
> -Uncomment the line in log_message to send messages to Sentry, however, be
> careful with this and avoid running too many tests with this on because we
> can easily go over the 5k events limit in Sentry. (We will probably need to
> add environment variables to control this and make sure it only runs in
> production. I haven't seen a way to do it in Sentry but it can be done with
> Python code).
> -I haven't found a good way to test some of the code, particularly in
> game.py. It should be safe and error free, but the main concern is making
> sure that all of the game attributes possible are sent to log_message and
> that the data is in the right format.
>
> Copy/pasting some info on the attributes in my email:
>
> Am I right in thinking that the logging function should be stateless? The
> reason that I'm asking is because a lot of attributes like game_id,
> player_id, phone_number, etc. are not available every time the logging
> function is called. They are only available in certain contexts.
>
> If we could save the state so that these attributes are saved the first
> time they are available, then it solves that problem, but from my
> understanding of the backend, the values would end up being incorrect and
> not accurately reflect the values of the user.
>
> IF we can't save the state, many times the logging function would not have
> those values, and I'm not sure how much value logging those messages would
> bring, because we wouldn't be able to easily pair the logging message with
> the user who caused the message. Possibly we might be able to figure it out
> from location or browser, but only for a very small number of users.
>
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/un7c0rn/stopthevirus/pull/132
> Commit Summary
>
>    - Add test logging to sentry
>    - Add game_id tag to sentry call
>    - Update senry project url
>    - Add phone_numbers, game_id, and player_id to logging and implement
>    in to_dict
>    - Add support for more attributes in logging for test_events
>    - Refactor logging and update engine and events for it
>    - Update all log_message calls to include game attributes
>    - Add comments
>
> File Changes
>
>    - *M* backend/game.py
>    <https://github.com/un7c0rn/stopthevirus/pull/132/files#diff-8193a9610414aac5611b5ca70ed29e9b>
>    (94)
>    - *M* backend/game_engine/common.py
>    <https://github.com/un7c0rn/stopthevirus/pull/132/files#diff-7ac9d9df200f224546f92acd82c1e0c4>
>    (78)
>    - *M* backend/game_engine/engine.py
>    <https://github.com/un7c0rn/stopthevirus/pull/132/files#diff-625be706e8e09f4da85165e3b95d5998>
>    (12)
>    - *M* backend/game_engine/events.py
>    <https://github.com/un7c0rn/stopthevirus/pull/132/files#diff-18b5463ed5ebd510409999ab18deacea>
>    (9)
>    - *M* backend/game_engine/firestore.py
>    <https://github.com/un7c0rn/stopthevirus/pull/132/files#diff-85feed5717383933103b890c89caad56>
>    (3)
>    - *M* backend/game_engine/twilio.py
>    <https://github.com/un7c0rn/stopthevirus/pull/132/files#diff-451c3b5bc4e748efa0b5922eb2f012ab>
>    (2)
>    - *M* requirements.txt
>    <https://github.com/un7c0rn/stopthevirus/pull/132/files#diff-b4ef698db8ca845e5845c4618278f29a>
>    (3)
>
> Patch Links:
>
>    - https://github.com/un7c0rn/stopthevirus/pull/132.patch
>    - https://github.com/un7c0rn/stopthevirus/pull/132.diff
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/un7c0rn/stopthevirus/pull/132>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADHNTTESRAJY3EG7MQJPBCLRRKYP3ANCNFSM4M72N4UQ>
> .
>

un7c0rn commented on backend/game_engine/common.py:
```diff
@@ -211,5 +224,64 @@ def to_json(self):
 
-def log_message(message):
-    logging.info(message)
+def log_message(message, game_attributes):
```
Let's make log_message stateless and explicitly define the mandatory and optional arguments for the function. For example:

log_message( message: Text, game_id: Text = None, additional_tags: Dict = None, push_to_sentry=False)

In this case message is a mandatory argument. Game ID may be optionally supplied. 'additional_tags' is a user defined dict of any metadata that should be logged.

un7c0rn commented on backend/game_engine/common.py:
```diff
@@ -211,5 +224,64 @@ def to_json(self):
+    tribe_id=""
+    game_id=""
+    player_id=""
+    if game_attributes:
```
By using a dict as the type for additional tags, we can simplify this code by iterating over all key value pairs in the additional tags dict and calling scope.set_tag accordingly. This puts more work on the caller to supply the tags that they want to be logged, but I think this is a good thing.

un7c0rn commented on backend/game_engine/common.py:
```diff
@@ -211,5 +224,64 @@ def to_json(self):
+            scope.set_tag("player_id", player_id)
+
+        if tiktok:
+            print("tiktok "+tiktok)
```
nit: add spaces for + operator.

un7c0rn commented on backend/game_engine/common.py:
```diff
@@ -211,5 +224,64 @@ def to_json(self):
+                print("phone_number " + phone_numbers)
+
+        if game_id:
+            print("game_id "+game_id)
```
nit: replace all print statements with logging.info(...) before merging.

un7c0rn commented on backend/game_engine/common.py:
```diff
@@ -211,5 +224,64 @@ def to_json(self):
+            print("team_id "+team_id)
+            scope.set_tag("team_id", team_id)
+
+        #UNCOMMENT THE BELOW LINE TO ENABLE SENTRY CALLS
```
With the updated method prototype, the push_to_sentry argument (or equivalent) can be default false, and used to conditionally perform remote logging. Then we can uncomment this line and switch the default to True for production.

If push_to_sentry is False, then log_message should simply call logging.info(...) with the formatted message text.

un7c0rn reviewed (changes requested):
Overall looks good. Minor changes requested inline. Can you also attach a screenshot of what the logged messages look like in Sentry?

Thanks!
Brandon

dzx11 commented:
@un7c0rn Here are a few screenshots. We can search/filter by any of the tags. (The issues are crossed out because I marked them as resolved)
![semtry1](https://user-images.githubusercontent.com/13459787/81899224-4972ae00-95f5-11ea-8f33-e4151c1514f9.PNG)
![sentry2](https://user-images.githubusercontent.com/13459787/81899227-4aa3db00-95f5-11ea-8c00-da5de2b9c169.PNG)

bernardbaker commented:
That looks great!

That will really benefit us in production.

On Thu, 14 May 2020, 07:12 dzx11, <notifications@github.com> wrote:

> @un7c0rn <https://github.com/un7c0rn> Here are a few screenshots. We can
> search/filter by any of the tags
> [image: semtry1]
> <https://user-images.githubusercontent.com/13459787/81899224-4972ae00-95f5-11ea-8f33-e4151c1514f9.PNG>
> [image: sentry2]
> <https://user-images.githubusercontent.com/13459787/81899227-4aa3db00-95f5-11ea-8c00-da5de2b9c169.PNG>
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/un7c0rn/stopthevirus/pull/132#issuecomment-628410355>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADHNTTGUK6XFVIFRD2TSOE3RRODWTANCNFSM4M72N4UQ>
> .
>

dzx11 committed:
Merge branch 'develop' into feature/sentry

dzx11 committed:
Refactor so logging does less data parsing

dzx11 committed:
Change sentry init to only init if push_to_sentry is true; clean up code

dzx11 commented:
@un7c0rn I've refactored this. I also found out that after Sentry is initialized, it will automatically log errors to the remote server without explicitly making calls (of course). This is probably undesirable in dev so I changed the code to only initialize the sdk on a truthy push_to_sentry value. Later we can change this to just initialize the sdk early in the application if we are in production.

un7c0rn commented on backend/game.py:
```diff
@@ -42,79 +42,89 @@ def __init__(self, game_id: Text, options: GameOptions):
         self._game_id = game_id
 
     def play(self, tribe1: Tribe, tribe2: Tribe, gamedb: Database, engine: Engine) -> Player:
+        game_id = ""
```
since _game_id is set in the Game constructor we can omit this check and just use self._game_id in the call to log_message.

un7c0rn commented on backend/game.py:
```diff
@@ -42,79 +42,89 @@ def __init__(self, game_id: Text, options: GameOptions):
         last_tribe_standing = self._play_multi_tribe(tribe1=tribe1, tribe2=tribe2,
                                                      gamedb=gamedb, engine=engine)
-        log_message("Last tribe standing is {}.".format(last_tribe_standing))
+        log_message("Last tribe standing is {}.".format(last_tribe_standing), game_id)
```
use named arguments for readability i.e.

log_message(message="...", game_id="...")

un7c0rn commented on backend/game_engine/events.py:
```diff
@@ -105,36 +105,42 @@ def _delete_message(self, receipt_handle: Text) -> None:
                                     ReceiptHandle=receipt_handle)
 
     def get(self) -> SMSEvent:
+        game_id = ""
```
in cases where we'd like to log messages from objects which do not have a game_id explicitly set in the constructor, we may need to modify the constructor to accept a game_id and ensure that it's populated.

un7c0rn commented on backend/game_engine/common.py:
```diff
@@ -187,7 +194,18 @@ def _to_dict_item(self, v):
-        log_message('to_dict called for {}'.format(self.__class__))
+        def massage_for_logging(attrs_dict):
+            if hasattr(self, 'recipient_phone_numbers'):
+                #sometimes phone_numbers is not a list
```
nit: please format comments in accordance with PEP 8 style guide

https://www.python.org/dev/peps/pep-0008/#block-comments

un7c0rn commented on backend/game_engine/common.py:
```diff
@@ -187,7 +194,18 @@ def _to_dict_item(self, v):
 
     def to_dict(self):
-        log_message('to_dict called for {}'.format(self.__class__))
+        def massage_for_logging(attrs_dict):
```
I think the call to log_message within the to_dict method will be too expensive and we can likely remove the call entirely here to simplify this. I initially had it here for debugging during the initial development.

un7c0rn reviewed (changes requested):
Screenshots and refactoring looks great super excited about getting this feature merged. Minor notes inline.

Thanks!
Brandon

dzx11 committed:
Refactor to explicitly include game_id

dzx11 commented:
@un7c0rn that makes sense. I've refactored so the game_id is explicitly set in all current log_message calls. PTAL, thanks

un7c0rn reviewed (approved):
LGTM!

un7c0rn committed:
Merge branch 'develop' into feature/sentry

un7c0rn commented:
@dzx11 I think we can merge this.

dzx11 committed:
Merge branch 'develop' into feature/sentry
